src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp

```diff
@@ -1275,6 +1275,9 @@ address TemplateInterpreterGenerator::generate_native_entry(bool synchronized) {
   // jvmti support
   __ notify_method_entry();
 
+  // AOT training run support
+  __ end_training_check();
+
```

Comment:
I wonder how much this will slow down the interpreter, as we trap into the VM on every interpreter entry. Could you do something like

```
long start = System.currentTimeMillis();
for (int i = 0; i < 10000000; i++) {
    emptyMethod();
}
System.out.println(System.currentMillis() - start();

```
and run it with something like this? 

```
rm foo.cds
java -Xint  -XX:+UnlockDiagnosticVMOptions -XX:+CDSManualFinalImage -XX:CacheDataStore=foo.cds ...
```

Reply:
The latest commit changes how we add triggers to the interpreter.  I added variants for the prologs that can have triggers, so only flagged methods with trap in the VM.  Also as the triggers are now only in flagged methods the VM code is faster as it know longer has to wrangle the method handle to test the flags.  While the traps remain even after the end of training has been triggered, the early out tests reduce the performance impact